// @stoqey/ibkr/dist/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ibkrPositionTossPosition = exports.OrderType = exports.OrderAction = void 0;
var instrument_utils_1 = require("./utils/instrument.utils");
;
;
var OrderAction;
(function (OrderAction) {
    OrderAction["BUY"] = "BUY";
    OrderAction["SELL"] = "SELL";
})(OrderAction || (exports.OrderAction = OrderAction = {}));
var OrderType;
(function (OrderType) {
    OrderType["MARKET"] = "MARKET";
    OrderType["LIMIT"] = "LIMIT";
    OrderType["STOP"] = "STOP";
    OrderType["STOP_LIMIT"] = "STOP_LIMIT";
})(OrderType || (exports.OrderType = OrderType = {}));
var ibkrPositionTossPosition = function (position) {
    var contract = position.contract;
    var symbol = (0, instrument_utils_1.getSymbolKey)(contract);
    var newPosition = {
        id: symbol,
        instrument: contract,
        action: position.pos > 0 ? OrderAction.BUY : OrderAction.SELL,
        quantity: Math.abs(position.pos),
        price: position.avgCost,
        lastPrice: position.marketPrice,
        // TODO date, entryDate
    };
    return newPosition;
};
exports.ibkrPositionTossPosition = ibkrPositionTossPosition;
