// @stoqey/ibkr/dist/marketdata/MarketDataManager.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarketDataManager = void 0;
var moment_1 = __importDefault(require("moment"));
var rxjs_1 = require("rxjs");
var IBKRConnection_1 = __importDefault(require("../connection/IBKRConnection"));
var ib_1 = require("@stoqey/ib");
var awaitP_1 = __importDefault(require("../utils/awaitP"));
var log_1 = require("../utils/log");
var events_1 = require("../events");
var Portfolios_1 = __importDefault(require("../portfolios/Portfolios"));
var isEmpty_1 = __importDefault(require("lodash/isEmpty"));
var time_utils_1 = require("../utils/time.utils");
var instrument_utils_1 = require("../utils/instrument.utils");
var data_utils_1 = require("../utils/data.utils");
var chart_1 = require("../utils/chart");
var sortBy_1 = __importDefault(require("lodash/sortBy"));
var utils_1 = require("../utils");
var Mkd_1 = require("./Mkd");
var appEvents = events_1.IBKREvents.Instance;
var logsNames = 'MkdMgr';
var MarketDataManager = /** @class */ (function (_super) {
    __extends(MarketDataManager, _super);
    function MarketDataManager() {
        var _this = _super.call(this) || this;
        _this.logsNames = logsNames;
        _this.GetHistoricalDataUpdates = new Map();
        _this.GetTickByTickDataUpdates = new Map();
        _this.currentBarData = new Map();
        _this.currentTickBarData = new Map();
        _this.removeHistoricalDataUpdates = function (contract) {
            var symbolId = _this.getSymbolKey(contract);
            var subscription = _this.GetHistoricalDataUpdates.get(symbolId);
            if (subscription) {
                subscription.unsubscribe();
                _this.GetHistoricalDataUpdates.delete(symbolId);
            }
            _this.currentBarData.delete(symbolId);
        };
        _this.getHistoricalDataUpdates = function (contract, barSizeSetting, whatToShow) { return __awaiter(_this, void 0, void 0, function () {
            var portfoliosManager_1, symbolId_1, getMarketDataLast30Minutes, e_1;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        portfoliosManager_1 = Portfolios_1.default.Instance;
                        if (!(!contract.conId || !contract.exchange)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.getContract(contract)];
                    case 1:
                        contract = _a.sent();
                        _a.label = 2;
                    case 2:
                        symbolId_1 = this.getSymbolKey(contract);
                        if (this.GetHistoricalDataUpdates.has(symbolId_1)) {
                            (0, log_1.warn)("".concat(logsNames, ".getHistoricalDataUpdates"), "Already subscribed to ".concat(symbolId_1));
                            return [2 /*return*/];
                        }
                        getMarketDataLast30Minutes = function () { return __awaiter(_this, void 0, void 0, function () {
                            var endDateTime, durationStr, barSizeSetting5Sec, historicalData, first, last, lastMarketData;
                            var _this = this;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        (0, log_1.log)("".concat(logsNames, ".getHistoricalDataUpdates.getMarketDataLast30Minutes"), "".concat(symbolId_1));
                                        endDateTime = (0, moment_1.default)().format('YYYYMMDD HH:mm:ss');
                                        durationStr = '3600 S';
                                        barSizeSetting5Sec = '5 secs';
                                        return [4 /*yield*/, (0, awaitP_1.default)(this.getHistoricalData(contract, endDateTime, durationStr, barSizeSetting5Sec, whatToShow))];
                                    case 1:
                                        historicalData = (_a.sent())[0];
                                        if (!(0, isEmpty_1.default)(historicalData)) {
                                            first = historicalData && historicalData[0];
                                            last = (historicalData || [])[historicalData.length - 1];
                                            (0, chart_1.plotMkdCli)(historicalData);
                                            (0, log_1.log)("".concat(logsNames, ".getHistoricalDataUpdates.getMarketDataLast30Minutes.length"), "".concat(historicalData === null || historicalData === void 0 ? void 0 : historicalData.length, " data points for ").concat(symbolId_1, " from ").concat((0, time_utils_1.formatDateStr)(first === null || first === void 0 ? void 0 : first.date), " to ").concat((0, time_utils_1.formatDateStr)(last === null || last === void 0 ? void 0 : last.date), " @").concat((0, data_utils_1.formatDec)(first === null || first === void 0 ? void 0 : first.close), " -> @").concat((0, data_utils_1.formatDec)(last === null || last === void 0 ? void 0 : last.close), " whatToShow=").concat(whatToShow));
                                            historicalData.forEach(function (marketDataItem) {
                                                _this.cacheBar(marketDataItem);
                                            });
                                            lastMarketData = historicalData[historicalData.length - 1];
                                            if (lastMarketData) {
                                                portfoliosManager_1.updateMarketPrice(contract.conId, lastMarketData.close);
                                            }
                                        }
                                        return [2 /*return*/];
                                }
                            });
                        }); };
                        return [4 /*yield*/, getMarketDataLast30Minutes()];
                    case 3:
                        _a.sent();
                        (0, log_1.log)("".concat(logsNames, ".getHistoricalDataUpdates"), "Subscribing to ".concat(symbolId_1));
                        this.GetHistoricalDataUpdates.set(symbolId_1, IBKRConnection_1.default.Instance.ib.getHistoricalDataUpdates(contract, barSizeSetting, whatToShow, 2)
                            .pipe((0, rxjs_1.catchError)(function (error) {
                            (0, log_1.warn)("".concat(logsNames, ".getHistoricalDataUpdates"), "Error subscribing to ".concat(symbolId_1), error);
                            _this.GetHistoricalDataUpdates.delete(symbolId_1);
                            return (0, rxjs_1.of)(null);
                        }))
                            .subscribe(function (bar) {
                            if (!bar)
                                return;
                            var currentTime = new Date(); // Use current time for consistent bar-to-bar timing
                            var barTime = +bar.time * 1000; // IBKR's bar timestamp
                            var barMinute = new Date(currentTime).setSeconds(0, 0); // use current time since bigger intervals send same timestamp for each bar
                            var prevBarData = _this.currentBarData.get(symbolId_1);
                            var incrementalVolume = bar.volume;
                            if (prevBarData && (prevBarData === null || prevBarData === void 0 ? void 0 : prevBarData.barMinute) === barMinute) {
                                incrementalVolume = Math.max(0, bar.volume - prevBarData.cumulativeVolume);
                            }
                            _this.currentBarData.set(symbolId_1, {
                                barTime: barTime,
                                cumulativeVolume: bar.volume,
                                barMinute: barMinute
                            });
                            var marketDataItem = {
                                instrument: contract,
                                date: currentTime, // Use current time instead of bar.time
                                open: bar.open,
                                high: bar.high,
                                low: bar.low,
                                close: bar.close,
                                volume: incrementalVolume, // Use incremental volume
                                wap: bar.WAP,
                                vwap: bar.WAP,
                                count: bar === null || bar === void 0 ? void 0 : bar.count,
                            };
                            _this.cacheBar(marketDataItem);
                            if (bar === null || bar === void 0 ? void 0 : bar.close) {
                                portfoliosManager_1.updateMarketPrice(contract.conId, bar.close);
                            }
                            appEvents.emit(events_1.IBKREVENTS.IBKR_BAR, marketDataItem);
                            var volumeStr = incrementalVolume > 0 ? "vol=".concat(incrementalVolume) : '';
                            var cumulativeStr = bar.volume !== incrementalVolume ? "cumVol=".concat(bar.volume) : '';
                            var isNewBar = !prevBarData || prevBarData.barMinute !== barMinute;
                            var barStatus = isNewBar ? '[NEW]' : '[UPD]';
                            (0, log_1.log)("".concat(logsNames, ".HDU"), "bar for ".concat(symbolId_1, " at ").concat((0, time_utils_1.formatDateStr)(currentTime), " @").concat(bar.close, " ").concat(volumeStr, " ").concat(cumulativeStr, " ").concat(barStatus).trim());
                        }));
                        return [3 /*break*/, 5];
                    case 4:
                        e_1 = _a.sent();
                        (0, log_1.warn)("getHistoricalDataUpdates error", e_1);
                        return [2 /*return*/];
                    case 5: return [2 /*return*/];
                }
            });
        }); };
        _this.getHistoricalData = function (contract_1, endDateTime_1, durationStr_1, barSizeSetting_1, whatToShow_1) {
            var args_1 = [];
            for (var _i = 5; _i < arguments.length; _i++) {
                args_1[_i - 5] = arguments[_i];
            }
            return __awaiter(_this, __spreadArray([contract_1, endDateTime_1, durationStr_1, barSizeSetting_1, whatToShow_1], args_1, true), void 0, function (contract, endDateTime, durationStr, barSizeSetting, whatToShow, useRTH) {
                var _a, contractInstrument, errContract, symbol, _b, bars, err, mkd;
                if (useRTH === void 0) { useRTH = false; }
                return __generator(this, function (_c) {
                    switch (_c.label) {
                        case 0: return [4 /*yield*/, (0, awaitP_1.default)(this.getContract(contract))];
                        case 1:
                            _a = _c.sent(), contractInstrument = _a[0], errContract = _a[1];
                            if (errContract) {
                                (0, log_1.warn)("getHistoricalData contract err", errContract);
                                return [2 /*return*/, null];
                            }
                            if (!contractInstrument) {
                                (0, log_1.warn)("getHistoricalData contract not found", contract);
                                return [2 /*return*/, null];
                            }
                            symbol = this.getSymbolKey(contractInstrument);
                            return [4 /*yield*/, (0, awaitP_1.default)(IBKRConnection_1.default.Instance.ib.getHistoricalData(contractInstrument, endDateTime, durationStr, barSizeSetting, whatToShow, useRTH, 2))];
                        case 2:
                            _b = _c.sent(), bars = _b[0], err = _b[1];
                            if (bars && bars.length > 0) {
                                mkd = bars.map(function (bar) {
                                    var date = new Date(+bar.time * 1000);
                                    var marketDataItem = {
                                        instrument: contractInstrument === null || contractInstrument === void 0 ? void 0 : contractInstrument.contract,
                                        date: date,
                                        open: bar.open,
                                        high: bar.high,
                                        low: bar.low,
                                        close: bar.close,
                                        volume: bar.volume,
                                        wap: bar.WAP,
                                        vwap: bar.WAP, // same as wap
                                        count: bar === null || bar === void 0 ? void 0 : bar.count,
                                    };
                                    return marketDataItem;
                                });
                                return [2 /*return*/, (0, sortBy_1.default)(mkd, 'date')];
                            }
                            if (err) {
                                (0, log_1.warn)("getHistoricalData err ".concat(symbol), err);
                            }
                            return [2 /*return*/, null];
                    }
                });
            });
        };
        _this.getHistoricalTicksLast = function (contract_1, startDate_1, endDate_1) {
            var args_1 = [];
            for (var _i = 3; _i < arguments.length; _i++) {
                args_1[_i - 3] = arguments[_i];
            }
            return __awaiter(_this, __spreadArray([contract_1, startDate_1, endDate_1], args_1, true), void 0, function (contract, startDate, endDate, numberOfTicks, useRTH) {
                var startDateTime, endDateTime, _a, contractInstrument, errContract, symbol, _b, ticks, err, mkd;
                if (numberOfTicks === void 0) { numberOfTicks = 1000; }
                if (useRTH === void 0) { useRTH = false; }
                return __generator(this, function (_c) {
                    switch (_c.label) {
                        case 0:
                            if (!startDate) {
                                (0, log_1.warn)("getHistoricalTicksLast startDate not set");
                                return [2 /*return*/, null];
                            }
                            if (endDate && startDate > endDate) {
                                (0, log_1.warn)("getHistoricalTicksLast startDate cannot be great than endDate");
                                return [2 /*return*/, null];
                            }
                            startDateTime = typeof startDate === 'string' ? startDate : (0, moment_1.default)(startDate).format('YYYYMMDD HH:mm:ss');
                            endDateTime = typeof endDate === 'string' ? endDate : (0, moment_1.default)(endDate).format('YYYYMMDD HH:mm:ss');
                            return [4 /*yield*/, (0, awaitP_1.default)(this.getContract(contract))];
                        case 1:
                            _a = _c.sent(), contractInstrument = _a[0], errContract = _a[1];
                            if (errContract) {
                                (0, log_1.warn)("getHistoricalTicksLast contract err", errContract);
                                return [2 /*return*/, null];
                            }
                            if (!contractInstrument) {
                                (0, log_1.warn)("getHistoricalTicksLast contract not found", contract);
                                return [2 /*return*/, null];
                            }
                            symbol = this.getSymbolKey(contractInstrument);
                            return [4 /*yield*/, (0, awaitP_1.default)((0, rxjs_1.lastValueFrom)(IBKRConnection_1.default.Instance.ib.getHistoricalTicksLast(contractInstrument, startDateTime, endDateTime, numberOfTicks, useRTH)))];
                        case 2:
                            _b = _c.sent(), ticks = _b[0], err = _b[1];
                            if (ticks && ticks.length > 0) {
                                mkd = ticks.map(function (bar) {
                                    var date = new Date(+bar.time * 1000);
                                    var tickDataItem = {
                                        contract: contractInstrument === null || contractInstrument === void 0 ? void 0 : contractInstrument.contract,
                                        date: date,
                                        price: bar.price,
                                        size: bar.size,
                                        exchange: bar.exchange,
                                        specialConditions: bar.specialConditions,
                                    };
                                    return tickDataItem;
                                });
                                return [2 /*return*/, (0, sortBy_1.default)(mkd, 'date')];
                            }
                            if (err) {
                                (0, log_1.warn)("getHistoricalTicksLast err ".concat(symbol), err);
                            }
                            return [2 /*return*/, null];
                    }
                });
            });
        };
        _this.getTickByTickDataUpdates = function (contract) { return __awaiter(_this, void 0, void 0, function () {
            var portfoliosManager_2, contractInstrument, symbolId_2, getMarketDataLast30Minutes, e_2;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        portfoliosManager_2 = Portfolios_1.default.Instance;
                        if (!(!contract.conId || !contract.exchange)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.getContract(contract)];
                    case 1:
                        contractInstrument = (_a.sent());
                        if (!contractInstrument) {
                            (0, log_1.warn)("".concat(logsNames, ".getTickByTickDataUpdates"), "Contract not found ".concat(contract));
                            return [2 /*return*/];
                        }
                        contract = contractInstrument.contract;
                        _a.label = 2;
                    case 2:
                        symbolId_2 = (0, instrument_utils_1.getSymbolKey)(contract);
                        if (this.GetTickByTickDataUpdates.has(symbolId_2)) {
                            (0, log_1.warn)("".concat(logsNames, ".getTickByTickDataUpdates"), "Already subscribed to ".concat(symbolId_2));
                            return [2 /*return*/];
                        }
                        getMarketDataLast30Minutes = function () { return __awaiter(_this, void 0, void 0, function () {
                            var endDateTime, durationStr, barSizeSetting5Sec, historicalData, first, last, lastMarketData;
                            var _this = this;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        (0, log_1.log)("".concat(logsNames, ".getTickByTickDataUpdates.getMarketDataLast30Minutes"), "".concat(symbolId_2));
                                        endDateTime = (0, moment_1.default)().format('YYYYMMDD HH:mm:ss');
                                        durationStr = '3600 S';
                                        barSizeSetting5Sec = '5 secs';
                                        return [4 /*yield*/, (0, awaitP_1.default)(this.getHistoricalData(contract, endDateTime, durationStr, barSizeSetting5Sec, ib_1.WhatToShow.TRADES))];
                                    case 1:
                                        historicalData = (_a.sent())[0];
                                        if (!(0, isEmpty_1.default)(historicalData)) {
                                            first = historicalData && historicalData[0];
                                            last = (historicalData || [])[historicalData.length - 1];
                                            (0, chart_1.plotMkdCli)(historicalData);
                                            (0, log_1.log)("".concat(logsNames, ".getTickByTickDataUpdates.getMarketDataLast30Minutes.length"), "".concat(historicalData === null || historicalData === void 0 ? void 0 : historicalData.length, " data points for ").concat(symbolId_2, " from ").concat((0, time_utils_1.formatDateStr)(first === null || first === void 0 ? void 0 : first.date), " to ").concat((0, time_utils_1.formatDateStr)(last === null || last === void 0 ? void 0 : last.date), " @").concat((0, data_utils_1.formatDec)(first === null || first === void 0 ? void 0 : first.close), " -> @").concat((0, data_utils_1.formatDec)(last === null || last === void 0 ? void 0 : last.close), " whatToShow=TRADES"));
                                            historicalData.forEach(function (marketDataItem) {
                                                _this.cacheBar(marketDataItem);
                                            });
                                            lastMarketData = historicalData[historicalData.length - 1];
                                            if (lastMarketData) {
                                                portfoliosManager_2.updateMarketPrice(contract.conId, lastMarketData.close);
                                            }
                                        }
                                        return [2 /*return*/];
                                }
                            });
                        }); };
                        return [4 /*yield*/, getMarketDataLast30Minutes()];
                    case 3:
                        _a.sent();
                        (0, log_1.log)("".concat(logsNames, ".getTickByTickDataUpdates"), "Subscribing to ".concat(symbolId_2));
                        this.GetTickByTickDataUpdates.set(symbolId_2, this.ib.getTickByTickAllLastDataUpdates(contract, 0, false)
                            .pipe((0, rxjs_1.catchError)(function (error) {
                            (0, log_1.warn)("".concat(logsNames, ".getTickByTickDataUpdates"), "Error subscribing to ".concat(symbolId_2), error);
                            _this.GetTickByTickDataUpdates.delete(symbolId_2);
                            return (0, rxjs_1.of)(null);
                        }))
                            .subscribe(function (tick) {
                            var tickData = {
                                contract: contract,
                                date: new Date(+tick.time * 1000),
                                price: tick.price,
                                size: tick.size,
                                exchange: tick.exchange,
                                specialConditions: tick.specialConditions,
                            };
                            _this.onTickByTickDataUpdates(tickData);
                            // log(`${logsNames}.TDU`, `tick for ${symbolId} at ${formatDateStr(new Date(tickData.date))} @${tickData.price} size=${tickData.size}  ${tickData.exchange ? `exchange=${tickData.exchange}` : ''} ${tickData.specialConditions ? `specialConditions=${tickData.specialConditions}` : ''}`);
                        }));
                        (0, log_1.log)("".concat(logsNames, ".getTickByTickDataUpdates"), "Subscribed to ".concat(symbolId_2));
                        return [3 /*break*/, 5];
                    case 4:
                        e_2 = _a.sent();
                        (0, log_1.warn)("getTickByTickDataUpdates error", e_2);
                        return [2 /*return*/];
                    case 5: return [2 /*return*/];
                }
            });
        }); };
        _this.onTickByTickDataUpdates = function (tick) {
            var symbolKey = (0, instrument_utils_1.getSymbolKey)(tick.contract);
            var tickDate = tick.date;
            var tickSecond = new Date(tickDate.setMilliseconds(0));
            var currentBarData = _this.currentTickBarData.get(symbolKey);
            // If we have existing data for this symbol
            if (currentBarData) {
                if (currentBarData.barSecond === tickSecond.getTime()) {
                    // SAME SECOND: Update the existing bar
                    currentBarData.close = tick.price;
                    currentBarData.volume += (tick.size || 0);
                    currentBarData.count++;
                    currentBarData.high = Math.max(currentBarData.high, tick.price);
                    currentBarData.low = Math.min(currentBarData.low, tick.price);
                    _this.currentTickBarData.set(symbolKey, currentBarData);
                }
                else {
                    // NEW SECOND: Send completed bar, start fresh
                    _this.cacheBar(currentBarData); // saves it just like the historical data updates
                    appEvents.emit(events_1.IBKREVENTS.IBKR_BAR, currentBarData);
                    (0, utils_1.logBar)("".concat(logsNames, ".TDU"), currentBarData);
                    // Create new bar for this tick
                    var newBar = {
                        instrument: tick.contract,
                        date: tickDate,
                        barTime: tickDate.getTime(),
                        barSecond: tickSecond.getTime(),
                        count: 1,
                        open: tick.price,
                        high: tick.price,
                        low: tick.price,
                        close: tick.price,
                        volume: tick.size
                    };
                    _this.currentTickBarData.set(symbolKey, newBar);
                }
            }
            else {
                // FIRST TICK: Create initial bar
                var newBar = {
                    instrument: tick.contract,
                    date: tickDate,
                    barTime: tickDate.getTime(),
                    barSecond: tickSecond.getTime(),
                    count: 1,
                    open: tick.price,
                    high: tick.price,
                    low: tick.price,
                    close: tick.price,
                    volume: tick.size
                };
                _this.currentTickBarData.set(symbolKey, newBar);
            }
            if (tick === null || tick === void 0 ? void 0 : tick.price) {
                Portfolios_1.default.Instance.updateMarketPrice(tick.contract.conId, tick.price);
            }
        };
        _this.removeTickByTickDataUpdates = function (contract) {
            var symbolId = _this.getSymbolKey(contract);
            var subscription = _this.GetTickByTickDataUpdates.get(symbolId);
            if (subscription) {
                subscription.unsubscribe();
                _this.GetTickByTickDataUpdates.delete(symbolId);
            }
            _this.currentTickBarData.delete(symbolId);
        };
        _this.getContract = function (contract) { return __awaiter(_this, void 0, void 0, function () {
            var _a, contracts, err, firstContract;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (contract === null || contract === void 0 ? void 0 : contract.contract) {
                            return [2 /*return*/, contract];
                        }
                        ;
                        return [4 /*yield*/, (0, awaitP_1.default)(this.ib.getContractDetails(contract))];
                    case 1:
                        _a = _b.sent(), contracts = _a[0], err = _a[1];
                        if (contracts && contracts.length > 0) {
                            firstContract = contracts[0];
                            return [2 /*return*/, __assign(__assign({}, firstContract), firstContract.contract)];
                        }
                        if (err) {
                            (0, log_1.warn)("getContract err ".concat(JSON.stringify(contract || {})), err);
                        }
                        return [2 /*return*/, null];
                }
            });
        }); };
        _this.searchContracts = function (contract) { return __awaiter(_this, void 0, void 0, function () {
            var _a, contracts, err;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, (0, awaitP_1.default)(this.ib.getContractDetails(contract))];
                    case 1:
                        _a = _b.sent(), contracts = _a[0], err = _a[1];
                        if (contracts && contracts.length > 0) {
                            return [2 /*return*/, contracts.map(function (c) {
                                    return __assign(__assign({}, c), c.contract);
                                })];
                        }
                        if (err) {
                            (0, log_1.warn)("getContract err ".concat(JSON.stringify(contract || {})), err);
                        }
                        return [2 /*return*/, []];
                }
            });
        }); };
        _this.init = function () {
            var ib = IBKRConnection_1.default.Instance.ib;
            if (!_this.ib) {
                _this.ib = ib;
                // ib.getHistoricalData // async
                // ib.getHistoricalDataUpdates() // subscribe
            }
        };
        return _this;
    }
    Object.defineProperty(MarketDataManager, "Instance", {
        get: function () {
            return this._instance || (this._instance = new this());
        },
        enumerable: false,
        configurable: true
    });
    return MarketDataManager;
}(Mkd_1.MkdManager));
exports.MarketDataManager = MarketDataManager;
exports.default = MarketDataManager;
