// @stoqey/ib/dist/core/io/enum/in-msg-id.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IN_MSG_ID = void 0;
/**
 * @internal
 *
 * Incoming message IDs.
 */
var IN_MSG_ID;
(function (IN_MSG_ID) {
    IN_MSG_ID[IN_MSG_ID["UNDEFINED"] = -1] = "UNDEFINED";
    IN_MSG_ID[IN_MSG_ID["TICK_PRICE"] = 1] = "TICK_PRICE";
    IN_MSG_ID[IN_MSG_ID["TICK_SIZE"] = 2] = "TICK_SIZE";
    IN_MSG_ID[IN_MSG_ID["ORDER_STATUS"] = 3] = "ORDER_STATUS";
    IN_MSG_ID[IN_MSG_ID["ERR_MSG"] = 4] = "ERR_MSG";
    IN_MSG_ID[IN_MSG_ID["OPEN_ORDER"] = 5] = "OPEN_ORDER";
    IN_MSG_ID[IN_MSG_ID["ACCT_VALUE"] = 6] = "ACCT_VALUE";
    IN_MSG_ID[IN_MSG_ID["PORTFOLIO_VALUE"] = 7] = "PORTFOLIO_VALUE";
    IN_MSG_ID[IN_MSG_ID["ACCT_UPDATE_TIME"] = 8] = "ACCT_UPDATE_TIME";
    IN_MSG_ID[IN_MSG_ID["NEXT_VALID_ID"] = 9] = "NEXT_VALID_ID";
    IN_MSG_ID[IN_MSG_ID["CONTRACT_DATA"] = 10] = "CONTRACT_DATA";
    IN_MSG_ID[IN_MSG_ID["EXECUTION_DATA"] = 11] = "EXECUTION_DATA";
    IN_MSG_ID[IN_MSG_ID["MARKET_DEPTH"] = 12] = "MARKET_DEPTH";
    IN_MSG_ID[IN_MSG_ID["MARKET_DEPTH_L2"] = 13] = "MARKET_DEPTH_L2";
    IN_MSG_ID[IN_MSG_ID["NEWS_BULLETINS"] = 14] = "NEWS_BULLETINS";
    IN_MSG_ID[IN_MSG_ID["MANAGED_ACCTS"] = 15] = "MANAGED_ACCTS";
    IN_MSG_ID[IN_MSG_ID["RECEIVE_FA"] = 16] = "RECEIVE_FA";
    IN_MSG_ID[IN_MSG_ID["HISTORICAL_DATA"] = 17] = "HISTORICAL_DATA";
    IN_MSG_ID[IN_MSG_ID["BOND_CONTRACT_DATA"] = 18] = "BOND_CONTRACT_DATA";
    IN_MSG_ID[IN_MSG_ID["SCANNER_PARAMETERS"] = 19] = "SCANNER_PARAMETERS";
    IN_MSG_ID[IN_MSG_ID["SCANNER_DATA"] = 20] = "SCANNER_DATA";
    IN_MSG_ID[IN_MSG_ID["TICK_OPTION_COMPUTATION"] = 21] = "TICK_OPTION_COMPUTATION";
    IN_MSG_ID[IN_MSG_ID["TICK_GENERIC"] = 45] = "TICK_GENERIC";
    IN_MSG_ID[IN_MSG_ID["TICK_STRING"] = 46] = "TICK_STRING";
    IN_MSG_ID[IN_MSG_ID["TICK_EFP"] = 47] = "TICK_EFP";
    IN_MSG_ID[IN_MSG_ID["CURRENT_TIME"] = 49] = "CURRENT_TIME";
    IN_MSG_ID[IN_MSG_ID["REAL_TIME_BARS"] = 50] = "REAL_TIME_BARS";
    IN_MSG_ID[IN_MSG_ID["FUNDAMENTAL_DATA"] = 51] = "FUNDAMENTAL_DATA";
    IN_MSG_ID[IN_MSG_ID["CONTRACT_DATA_END"] = 52] = "CONTRACT_DATA_END";
    IN_MSG_ID[IN_MSG_ID["OPEN_ORDER_END"] = 53] = "OPEN_ORDER_END";
    IN_MSG_ID[IN_MSG_ID["ACCT_DOWNLOAD_END"] = 54] = "ACCT_DOWNLOAD_END";
    IN_MSG_ID[IN_MSG_ID["EXECUTION_DATA_END"] = 55] = "EXECUTION_DATA_END";
    IN_MSG_ID[IN_MSG_ID["DELTA_NEUTRAL_VALIDATION"] = 56] = "DELTA_NEUTRAL_VALIDATION";
    IN_MSG_ID[IN_MSG_ID["TICK_SNAPSHOT_END"] = 57] = "TICK_SNAPSHOT_END";
    IN_MSG_ID[IN_MSG_ID["MARKET_DATA_TYPE"] = 58] = "MARKET_DATA_TYPE";
    IN_MSG_ID[IN_MSG_ID["COMMISSION_REPORT"] = 59] = "COMMISSION_REPORT";
    IN_MSG_ID[IN_MSG_ID["POSITION"] = 61] = "POSITION";
    IN_MSG_ID[IN_MSG_ID["POSITION_END"] = 62] = "POSITION_END";
    IN_MSG_ID[IN_MSG_ID["ACCOUNT_SUMMARY"] = 63] = "ACCOUNT_SUMMARY";
    IN_MSG_ID[IN_MSG_ID["ACCOUNT_SUMMARY_END"] = 64] = "ACCOUNT_SUMMARY_END";
    IN_MSG_ID[IN_MSG_ID["VERIFY_MESSAGE_API"] = 65] = "VERIFY_MESSAGE_API";
    IN_MSG_ID[IN_MSG_ID["VERIFY_COMPLETED"] = 66] = "VERIFY_COMPLETED";
    IN_MSG_ID[IN_MSG_ID["DISPLAY_GROUP_LIST"] = 67] = "DISPLAY_GROUP_LIST";
    IN_MSG_ID[IN_MSG_ID["DISPLAY_GROUP_UPDATED"] = 68] = "DISPLAY_GROUP_UPDATED";
    IN_MSG_ID[IN_MSG_ID["VERIFY_AND_AUTH_MESSAGE_API"] = 69] = "VERIFY_AND_AUTH_MESSAGE_API";
    IN_MSG_ID[IN_MSG_ID["VERIFY_AND_AUTH_COMPLETED"] = 70] = "VERIFY_AND_AUTH_COMPLETED";
    IN_MSG_ID[IN_MSG_ID["POSITION_MULTI"] = 71] = "POSITION_MULTI";
    IN_MSG_ID[IN_MSG_ID["POSITION_MULTI_END"] = 72] = "POSITION_MULTI_END";
    IN_MSG_ID[IN_MSG_ID["ACCOUNT_UPDATE_MULTI"] = 73] = "ACCOUNT_UPDATE_MULTI";
    IN_MSG_ID[IN_MSG_ID["ACCOUNT_UPDATE_MULTI_END"] = 74] = "ACCOUNT_UPDATE_MULTI_END";
    IN_MSG_ID[IN_MSG_ID["SECURITY_DEFINITION_OPTION_PARAMETER"] = 75] = "SECURITY_DEFINITION_OPTION_PARAMETER";
    IN_MSG_ID[IN_MSG_ID["SECURITY_DEFINITION_OPTION_PARAMETER_END"] = 76] = "SECURITY_DEFINITION_OPTION_PARAMETER_END";
    IN_MSG_ID[IN_MSG_ID["SOFT_DOLLAR_TIERS"] = 77] = "SOFT_DOLLAR_TIERS";
    IN_MSG_ID[IN_MSG_ID["FAMILY_CODES"] = 78] = "FAMILY_CODES";
    IN_MSG_ID[IN_MSG_ID["SYMBOL_SAMPLES"] = 79] = "SYMBOL_SAMPLES";
    IN_MSG_ID[IN_MSG_ID["MKT_DEPTH_EXCHANGES"] = 80] = "MKT_DEPTH_EXCHANGES";
    IN_MSG_ID[IN_MSG_ID["TICK_REQ_PARAMS"] = 81] = "TICK_REQ_PARAMS";
    IN_MSG_ID[IN_MSG_ID["SMART_COMPONENTS"] = 82] = "SMART_COMPONENTS";
    IN_MSG_ID[IN_MSG_ID["NEWS_ARTICLE"] = 83] = "NEWS_ARTICLE";
    IN_MSG_ID[IN_MSG_ID["TICK_NEWS"] = 84] = "TICK_NEWS";
    IN_MSG_ID[IN_MSG_ID["NEWS_PROVIDERS"] = 85] = "NEWS_PROVIDERS";
    IN_MSG_ID[IN_MSG_ID["HISTORICAL_NEWS"] = 86] = "HISTORICAL_NEWS";
    IN_MSG_ID[IN_MSG_ID["HISTORICAL_NEWS_END"] = 87] = "HISTORICAL_NEWS_END";
    IN_MSG_ID[IN_MSG_ID["HEAD_TIMESTAMP"] = 88] = "HEAD_TIMESTAMP";
    IN_MSG_ID[IN_MSG_ID["HISTOGRAM_DATA"] = 89] = "HISTOGRAM_DATA";
    IN_MSG_ID[IN_MSG_ID["HISTORICAL_DATA_UPDATE"] = 90] = "HISTORICAL_DATA_UPDATE";
    IN_MSG_ID[IN_MSG_ID["REROUTE_MKT_DATA"] = 91] = "REROUTE_MKT_DATA";
    IN_MSG_ID[IN_MSG_ID["REROUTE_MKT_DEPTH"] = 92] = "REROUTE_MKT_DEPTH";
    IN_MSG_ID[IN_MSG_ID["MARKET_RULE"] = 93] = "MARKET_RULE";
    IN_MSG_ID[IN_MSG_ID["PNL"] = 94] = "PNL";
    IN_MSG_ID[IN_MSG_ID["PNL_SINGLE"] = 95] = "PNL_SINGLE";
    IN_MSG_ID[IN_MSG_ID["HISTORICAL_TICKS"] = 96] = "HISTORICAL_TICKS";
    IN_MSG_ID[IN_MSG_ID["HISTORICAL_TICKS_BID_ASK"] = 97] = "HISTORICAL_TICKS_BID_ASK";
    IN_MSG_ID[IN_MSG_ID["HISTORICAL_TICKS_LAST"] = 98] = "HISTORICAL_TICKS_LAST";
    IN_MSG_ID[IN_MSG_ID["TICK_BY_TICK"] = 99] = "TICK_BY_TICK";
    IN_MSG_ID[IN_MSG_ID["ORDER_BOUND"] = 100] = "ORDER_BOUND";
    IN_MSG_ID[IN_MSG_ID["COMPLETED_ORDER"] = 101] = "COMPLETED_ORDER";
    IN_MSG_ID[IN_MSG_ID["COMPLETED_ORDERS_END"] = 102] = "COMPLETED_ORDERS_END";
    IN_MSG_ID[IN_MSG_ID["REPLACE_FA_END"] = 103] = "REPLACE_FA_END";
    IN_MSG_ID[IN_MSG_ID["WSH_META_DATA"] = 104] = "WSH_META_DATA";
    IN_MSG_ID[IN_MSG_ID["WSH_EVENT_DATA"] = 105] = "WSH_EVENT_DATA";
    IN_MSG_ID[IN_MSG_ID["HISTORICAL_SCHEDULE"] = 106] = "HISTORICAL_SCHEDULE";
    IN_MSG_ID[IN_MSG_ID["USER_INFO"] = 107] = "USER_INFO";
})(IN_MSG_ID || (exports.IN_MSG_ID = IN_MSG_ID = {}));
