// @stoqey/ibkr/dist/portfolios/Portfolios.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Portfolios = void 0;
var rxjs_1 = require("rxjs");
var ib_1 = require("@stoqey/ib");
var connection_1 = require("../connection");
var compact_1 = __importDefault(require("lodash/compact"));
var MarketDataManager_1 = __importDefault(require("../marketdata/MarketDataManager"));
var log_utils_1 = require("../utils/log.utils");
var log_1 = require("../utils/log");
var Portfolios = /** @class */ (function () {
    function Portfolios() {
        var _this = this;
        this.currentPortfolios = new Map();
        this.closedPositions = new Map();
        this.entryPrices = new Map();
        this.entryDates = new Map();
        this.updateMarketPrice = function (conId, close) {
            var _a;
            var position = (_a = _this.currentPortfolios) === null || _a === void 0 ? void 0 : _a.get(conId);
            if (position) {
                var newPosition = __assign({}, position);
                newPosition.marketPrice = close;
                _this.currentPortfolios.set(conId, newPosition);
            }
        };
        this.getEntryPrice = function (conId) {
            return _this.entryPrices.get(conId);
        };
        this.getEntryDate = function (conId) {
            return _this.entryDates.get(conId);
        };
        this.mapPositions = function (positionOg, subscribe) {
            var _a;
            if (subscribe === void 0) { subscribe = false; }
            if (!positionOg) {
                return;
            }
            var position = __assign({}, positionOg);
            var contractId = (_a = position === null || position === void 0 ? void 0 : position.contract) === null || _a === void 0 ? void 0 : _a.conId;
            if (contractId) {
                if (position.contract.exchange === "VALUE") {
                    return;
                }
                if (!position.pos) {
                    var positionToClose = _this.currentPortfolios.get(contractId);
                    if (positionToClose) {
                        _this.closedPositions.set(contractId, positionToClose);
                        _this.currentPortfolios.delete(contractId);
                        _this.entryPrices.delete(contractId);
                        _this.entryDates.delete(contractId);
                        (0, log_utils_1.logPosition)("Portfolios.mapPositions", positionToClose, true);
                    }
                }
                else {
                    // TODO: use this.entryDates or position.entryDate
                    position.entryDate = new Date();
                    var multiplier = position.contract.multiplier;
                    if (multiplier) {
                        position.avgCost = position.avgCost / multiplier;
                    }
                    _this.currentPortfolios.set(contractId, position);
                    _this.entryPrices.set(contractId, position.avgCost);
                    _this.entryDates.set(contractId, position.entryDate);
                    if (_this.closedPositions.has(contractId)) {
                        _this.closedPositions.delete(contractId);
                    }
                    // When successful
                    if (subscribe) {
                        var isBuy = position.pos > 0;
                        var whatToShow = isBuy ? ib_1.WhatToShow.ASK : ib_1.WhatToShow.BID;
                        var barSize = "5 secs";
                        MarketDataManager_1.default.Instance.getHistoricalDataUpdates(position.contract, barSize, whatToShow);
                    }
                    (0, log_utils_1.logPosition)("Portfolios.mapPositions", position);
                }
            }
            return position;
        };
        /**
         * getPortfolios
         */
        this.asyncPortfolios = function () { return __awaiter(_this, void 0, void 0, function () {
            var getPositions, positions;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, rxjs_1.firstValueFrom)(this.ib.getPositions())];
                    case 1:
                        getPositions = _a.sent();
                        positions = [];
                        getPositions.all.forEach(function (value, key) {
                            positions = (0, compact_1.default)(value.map(function (position) { return _this.mapPositions(position); }));
                        });
                        return [2 /*return*/, positions];
                }
            });
        }); };
        this.syncPortfolios = function () {
            MarketDataManager_1.default.Instance.init();
            _this.GetPositions = _this.ib.getPositions()
                .pipe((0, rxjs_1.catchError)(function (error) {
                (0, log_1.warn)("syncPortfolios", "Error subscribing to positions", error);
                return (0, rxjs_1.of)(null);
            }))
                .subscribe(function (accountUpdates) {
                accountUpdates.all.forEach(function (value, key) {
                    value.forEach(function (position) { return _this.mapPositions(position); });
                });
            });
        };
        this.disconnect = function () {
            _this.GetPositions.unsubscribe();
        };
        this.getPositions = function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.asyncPortfolios()];
            });
        }); };
        this.init = function () {
            if (!_this.ib) {
                _this.ib = connection_1.IBKRConnection.Instance.ib;
                _this.syncPortfolios();
            }
        };
    }
    Object.defineProperty(Portfolios, "Instance", {
        get: function () {
            return this._instance || (this._instance = new this());
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Portfolios.prototype, "positions", {
        get: function () {
            return Array.from(this.currentPortfolios.values());
        },
        enumerable: false,
        configurable: true
    });
    Portfolios.prototype.getLatestClosedPosition = function (conId) {
        return this.closedPositions.get(conId);
    };
    return Portfolios;
}());
exports.Portfolios = Portfolios;
exports.default = Portfolios;
