// @stoqey/ibkr/dist/connection/IBKRConnection.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IBKRConnection = void 0;
var ib_1 = require("@stoqey/ib");
var log_1 = require("../utils/log");
var Portfolios_1 = __importDefault(require("../portfolios/Portfolios"));
var Orders_1 = __importDefault(require("../orders/Orders"));
var MarketDataManager_1 = __importDefault(require("../marketdata/MarketDataManager"));
var events_1 = require("../events");
var AccountSummary_1 = require("../account/AccountSummary");
var ibkrEvents = events_1.IBKREvents.Instance;
var IBKRConnection = /** @class */ (function () {
    function IBKRConnection() {
        var _this = this;
        this.connected = false;
        /**
         * initializeDep
         * Call/Initialize Account summary -> Portfolios -> OpenOrders
         */
        this.initializeDep = function () { return __awaiter(_this, void 0, void 0, function () {
            var accountSummary, marketData, portfolio, openOrders, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 7, , 8]);
                        // 1. Account summary
                        (0, log_1.log)('1. Account summary');
                        accountSummary = AccountSummary_1.AccountSummary.Instance;
                        accountSummary.init();
                        return [4 /*yield*/, accountSummary.getAccountSummaryUpdates()];
                    case 1:
                        _a.sent();
                        // 2. Market data
                        (0, log_1.log)('2. Market data');
                        marketData = MarketDataManager_1.default.Instance;
                        return [4 /*yield*/, marketData.init()];
                    case 2:
                        _a.sent();
                        // 3. Portfolios
                        (0, log_1.log)('3. Portfolios');
                        portfolio = Portfolios_1.default.Instance;
                        return [4 /*yield*/, portfolio.init()];
                    case 3:
                        _a.sent();
                        return [4 /*yield*/, portfolio.asyncPortfolios()];
                    case 4:
                        _a.sent();
                        // 4. Orders
                        (0, log_1.log)('3. Orders');
                        openOrders = Orders_1.default.Instance;
                        return [4 /*yield*/, openOrders.init()];
                    case 5:
                        _a.sent();
                        return [4 /*yield*/, openOrders.asyncOpenOrders()];
                    case 6:
                        _a.sent();
                        return [2 /*return*/, true];
                    case 7:
                        error_1 = _a.sent();
                        (0, log_1.log)('error initialising IBKR', error_1);
                        return [2 /*return*/, false];
                    case 8: return [2 /*return*/];
                }
            });
        }); };
    }
    Object.defineProperty(IBKRConnection, "Instance", {
        get: function () {
            return this._instance || (this._instance = new this());
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(IBKRConnection.prototype, "ib", {
        get: function () {
            return this.ibApiNext;
        },
        enumerable: false,
        configurable: true
    });
    ;
    IBKRConnection.prototype.init = function (opt) {
        if (this.connected || this.ibApiNext) {
            (0, log_1.log)('already init');
            return Promise.resolve(true);
        }
        if (opt) {
            if (!opt.reconnectInterval)
                opt.reconnectInterval = 1000;
            if (!opt.connectionWatchdogInterval)
                opt.connectionWatchdogInterval = 1;
            (0, log_1.log)('init with options', opt);
            this.ibApiNext = new ib_1.IBApiNext(opt);
        }
        else {
            opt = {};
            // reconnect interval from env
            opt.reconnectInterval = parseInt(process.env.IBKR_RECONNECT_INTERVAL) || 1000;
            opt.connectionWatchdogInterval = parseInt(process.env.IBKR_WATCHDOG_INTERVAL) || 1;
            // port and host from env
            opt.host = process.env.IBKR_HOST || '127.0.0.1';
            opt.port = parseInt(process.env.IBKR_PORT) || 7497;
            (0, log_1.log)('init with env', opt);
            this.ibApiNext = new ib_1.IBApiNext(opt);
        }
        this.ibApiNext.logLevel = 5;
        return this.connect();
    };
    IBKRConnection.prototype.connect = function () {
        return __awaiter(this, arguments, void 0, function (clientId) {
            var _this = this;
            if (clientId === void 0) { clientId = 0; }
            return __generator(this, function (_a) {
                if (process.env.IBKR_CLIENT_ID) {
                    clientId = parseInt(process.env.IBKR_CLIENT_ID);
                }
                ;
                return [2 /*return*/, new Promise(function (resolve) {
                        // connect to IBKR
                        if (_this.connected) {
                            _this.initializeDep();
                            resolve(true);
                        }
                        _this.errors = _this.ibApiNext.errorSubject.subscribe(function (error) {
                            (0, log_1.log)("IBKRConnection.ibApiNext.errors", error.message);
                        });
                        _this.ibApiNext.connectionState.subscribe(function (state) {
                            (0, log_1.log)('connection state', state);
                            switch (state) {
                                case ib_1.ConnectionState.Connecting:
                                    _this.connectionState = state;
                                    (0, log_1.log)('connecting to ibkr');
                                    break;
                                case ib_1.ConnectionState.Connected:
                                    _this.connectionState = state;
                                    _this.connected = true;
                                    (0, log_1.log)('connected to ibkr');
                                    resolve(true);
                                    break;
                                case ib_1.ConnectionState.Disconnected:
                                    (0, log_1.log)('disconnected from ibkr', _this.connectionState);
                                    _this.connected = false;
                                    if (_this.connectionState === ib_1.ConnectionState.Connecting) {
                                        _this.connectionState = state;
                                        // if was connecting, then reject
                                        resolve(false);
                                    }
                                    break;
                                default:
                                    break;
                            }
                        });
                        if (!_this.connection) {
                            _this.connection = _this.ibApiNext.connectionState.subscribe(function (state) {
                                (0, log_1.log)('connection state', state);
                                switch (state) {
                                    case ib_1.ConnectionState.Connecting:
                                        _this.connectionState = state;
                                        (0, log_1.log)('connecting to ibkr');
                                        break;
                                    case ib_1.ConnectionState.Connected:
                                        _this.connectionState = state;
                                        _this.connected = true;
                                        _this.initializeDep();
                                        ibkrEvents.emit(events_1.IBKREVENTS.IBKR_CONNECTED);
                                        (0, log_1.log)('connected to ibkr');
                                        break;
                                    case ib_1.ConnectionState.Disconnected:
                                        (0, log_1.log)('disconnected from ibkr', _this.connectionState);
                                        _this.connected = false;
                                        // TODO disconnect
                                        if (_this.connectionState === ib_1.ConnectionState.Connecting) {
                                            _this.connectionState = state;
                                        }
                                        if (_this.connectionState === ib_1.ConnectionState.Disconnected) {
                                            // process.exit(0);
                                            console.log('ConnectionState.Disconnected', ib_1.ConnectionState.Disconnected);
                                        }
                                        break;
                                    default:
                                        break;
                                }
                            });
                        }
                        _this.ibApiNext.connect(clientId);
                    })];
            });
        });
    };
    IBKRConnection.prototype.disconnect = function () {
        // TODO others....
        if (this.connection) {
            this.connection.unsubscribe();
        }
    };
    return IBKRConnection;
}());
exports.IBKRConnection = IBKRConnection;
exports.default = IBKRConnection;
